File: backend/src/main/java/org/luxons/sevenwonders/validation/DestinationAccessValidator.java
package org.luxons.sevenwonders.validation;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.luxons.sevenwonders.lobby.Lobby;
import org.luxons.sevenwonders.repositories.LobbyRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class DestinationAccessValidator {

    private static final Pattern lobbyDestination = Pattern.compile(".*?/lobby/(?<id>\\d+?)(/.*)?");

    private static final Pattern gameDestination = Pattern.compile(".*?/game/(?<id>\\d+?)(/.*)?");

    private final LobbyRepository lobbyRepository;

    @Autowired
    public DestinationAccessValidator(LobbyRepository lobbyRepository) {
        this.lobbyRepository = lobbyRepository;
    }

    public boolean hasAccess(String username, String destination) {
        if (username == null) {
            // unnamed user cannot belong to anything
            return false;
        }
        if (hasForbiddenGameReference(username, destination)) {
            return false;
        }
        if (hasForbiddenLobbyReference(username, destination)) {
            return false;
        }
        return true;
    }

    private boolean hasForbiddenGameReference(String username, String destination) {
        Matcher gameMatcher = gameDestination.matcher(destination);
        if (!gameMatcher.matches()) {
            return false; // no game reference is always OK
        }
        int gameId = extractId(gameMatcher);
        return !isUserInLobby(username, gameId);
    }

    private boolean hasForbiddenLobbyReference(String username, String destination) {
        Matcher lobbyMatcher = lobbyDestination.matcher(destination);
        if (!lobbyMatcher.matches()) {
            return false; // no lobby reference is always OK
        }
        int lobbyId = extractId(lobbyMatcher);
        return !isUserInLobby(username, lobbyId);
    }

    private boolean isUserInLobby(String username, int lobbyId) {
        Lobby lobby = lobbyRepository.find(lobbyId);
        return lobby.containsUser(username);
    }

    private static int extractId(Matcher matcher) {
        String id = matcher.group("id");
        return Integer.parseInt(id);
    }
}

File: backend/src/main/kotlin/org/luxons/sevenwonders/validation/DestinationAccessValidator.kt
package org.luxons.sevenwonders.validation

import org.luxons.sevenwonders.repositories.LobbyRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.util.regex.Matcher
import java.util.regex.Pattern

@Component
class DestinationAccessValidator @Autowired constructor(private val lobbyRepository: LobbyRepository) {

    fun hasAccess(username: String?, destination: String): Boolean {
        return when {
            username == null -> false // unnamed user cannot belong to anything
            hasForbiddenGameReference(username, destination) -> false
            hasForbiddenLobbyReference(username, destination) -> false
            else -> true
        }
    }

    private fun hasForbiddenGameReference(username: String, destination: String): Boolean {
        val gameMatcher = gameDestination.matcher(destination)
        if (!gameMatcher.matches()) {
            return false // no game reference is always OK
        }
        val gameId = extractId(gameMatcher)
        return !isUserInLobby(username, gameId)
    }

    private fun hasForbiddenLobbyReference(username: String, destination: String): Boolean {
        val lobbyMatcher = lobbyDestination.matcher(destination)
        if (!lobbyMatcher.matches()) {
            return false // no lobby reference is always OK
        }
        val lobbyId = extractId(lobbyMatcher)
        return !isUserInLobby(username, lobbyId)
    }

    private fun isUserInLobby(username: String, lobbyId: Int): Boolean {
        val lobby = lobbyRepository.find(lobbyId.toLong())
        return lobby.containsUser(username)
    }

    companion object {

        private val lobbyDestination = Pattern.compile(".*?/lobby/(?<id>\\d+?)(/.*)?")

        private val gameDestination = Pattern.compile(".*?/game/(?<id>\\d+?)(/.*)?")

        private fun extractId(matcher: Matcher): Int {
            val id = matcher.group("id")
            return Integer.parseInt(id)
        }
    }
}
